geth-utils/gethutil/mpt/witness/leaf.go

```diff
@@ -567,3 +579,38 @@ func prepareStorageLeafNode(leafS, leafC, neighbourNode []byte, storage_key comm
 
 	return node
 }
+
+func equipLeafWithWrongExtension(leafNode Node, keyMiddle, keyAfter, nibblesMiddle, nibblesAfter []byte) Node {
+	l := len(leafNode.Values)
+	leafNode.Values[l-modifiedExtensionNodeRowLen] = keyMiddle
+	startNibblePos := 2 // we don't need any nibbles for case keyLen = 1
+	if len(keyMiddle) > 1 {
+		if len(nibblesMiddle)%2 == 0 {
+			startNibblePos = 1
+		} else {
+			startNibblePos = 2
+		}
+	}
+	ind := 0
+	for j := startNibblePos; j < len(nibblesMiddle); j += 2 {
+		leafNode.Values[l-modifiedExtensionNodeRowLen+1][2+ind] = nibblesMiddle[j]
+		ind++
+	}
+
+	leafNode.Values[l-modifiedExtensionNodeRowLen+3] = keyAfter
+	startNibblePos = 2 // we don't need any nibbles for case keyLen = 1
+	if len(keyAfter) > 1 {
+		if len(nibblesAfter)%2 == 0 {
+			startNibblePos = 1
+		} else {
+			startNibblePos = 2
+		}
+	}
+	ind = 0
+	for j := startNibblePos; j < len(nibblesAfter); j += 2 {
+		leafNode.Values[l-modifiedExtensionNodeRowLen+4][2+ind] = nibblesAfter[j]
```

Comment:
The length of `nibblesMiddle` can be arbitrary, it is stored in the row `leafNode.Values[l-modifiedExtensionNodeRowLen+1]`.
`nibblesAfter` are stored in the row `leafNode.Values[l-modifiedExtensionNodeRowLen+4]`, so `+4` means into which row the nibbles are stored.

If it's confusing because only `leafNode.Values[l-modifiedExtensionNodeRowLen+1]` and `leafNode.Values[l-modifiedExtensionNodeRowLen+4]` are used (and not for example `leafNode.Values[l-modifiedExtensionNodeRowLen+2]`), it's because these rows are marked to constrain that the values are nibbles, see the following enum in `witness_row.rs`:

```
pub(crate) enum StorageRowType {
    KeyS,
    ValueS,
    KeyC,
    ValueC,
    Drifted,
    Wrong,
    LongExtNodeKey,
    LongExtNodeNibbles,
    LongExtNodeValue,
    ShortExtNodeKey,
    ShortExtNodeNibbles,
    ShortExtNodeValue,
    Address,
    Key,
    Count,
}
```

The rows where `nibblesMiddle` and `nibblesAfter` are stored correspond to `LongExtNodeNibbles` and `ShortExtNodeNibbles`.

Reply:
Oh, I misread it. `leafNode.Values` is a double array so you assign `nibblesMiddle` to row `l-modifiedExtensionNodeRowLen+1` and assign `nibblesAfter` to row `l-modifiedExtensionNodeRowLen+4`, right?

Reply:
Yes, exactly!